训练模型（raindrop的train数据集）（input、target）(800张图片)

In [ ]:
!python train_diffusion.py --config ./configs/raindrop_configs.yml

=> using device: cuda
=> creating denoising diffusion model
=> current epoch:  0
step: 40, loss: 2535.799316, time consumption: 0.178849
step: 80, loss: 1325.088745, time consumption: 0.133702
step: 120, loss: 1061.279297, time consumption: 0.116841
step: 160, loss: 819.617615, time consumption: 0.107770
=> current epoch:  1
step: 200, loss: 732.220154, time consumption: 0.081916
step: 240, loss: 812.471069, time consumption: 0.082520
step: 280, loss: 551.671814, time consumption: 0.082711
step: 320, loss: 711.518372, time consumption: 0.082582
=> current epoch:  2
step: 360, loss: 436.944336, time consumption: 0.084747
step: 400, loss: 626.912598, time consumption: 0.085922
step: 440, loss: 651.929382, time consumption: 0.084505
step: 480, loss: 426.851471, time consumption: 0.083632
=> current epoch:  3
step: 520, loss: 404.627045, time consumption: 0.081209
step: 560, loss: 464.112885, time consumption: 0.085953
step: 600, loss: 533.672974, time consumption: 0.095025
step: 640, loss

step: 5280, loss: 178.711792, time consumption: 0.107406
=> current epoch:  33
step: 5320, loss: 278.347992, time consumption: 0.085056
step: 5360, loss: 161.763489, time consumption: 0.097800
step: 5400, loss: 101.317001, time consumption: 0.101742
step: 5440, loss: 221.076904, time consumption: 0.098249
=> current epoch:  34
step: 5480, loss: 148.582230, time consumption: 0.080115
step: 5520, loss: 315.348785, time consumption: 0.089832
step: 5560, loss: 341.320618, time consumption: 0.093786
step: 5600, loss: 260.767578, time consumption: 0.096509
=> current epoch:  35
step: 5640, loss: 123.053116, time consumption: 0.107383
step: 5680, loss: 142.650742, time consumption: 0.108390
step: 5720, loss: 136.210739, time consumption: 0.107476
step: 5760, loss: 402.545746, time consumption: 0.105660
=> current epoch:  36
step: 5800, loss: 130.980637, time consumption: 0.097085
step: 5840, loss: 173.679489, time consumption: 0.092637
step: 5880, loss: 153.370239, time consumption: 0.092181


step: 10480, loss: 150.612625, time consumption: 0.097645
step: 10520, loss: 295.174591, time consumption: 0.099648
step: 10560, loss: 132.389938, time consumption: 0.100609
=> current epoch:  66
step: 10600, loss: 185.211349, time consumption: 0.088283
step: 10640, loss: 120.271233, time consumption: 0.101693
step: 10680, loss: 80.683159, time consumption: 0.099056
step: 10720, loss: 127.805054, time consumption: 0.095582
=> current epoch:  67
step: 10760, loss: 112.103020, time consumption: 0.097670
step: 10800, loss: 203.891464, time consumption: 0.106233
step: 10840, loss: 270.224274, time consumption: 0.100435
step: 10880, loss: 114.495506, time consumption: 0.098454
=> current epoch:  68
step: 10920, loss: 162.982300, time consumption: 0.109925
step: 10960, loss: 76.785988, time consumption: 0.124517
step: 11000, loss: 134.539612, time consumption: 0.109673
step: 11040, loss: 84.985916, time consumption: 0.116620
=> current epoch:  69
step: 11080, loss: 68.300591, time consumptio

step: 15760, loss: 133.856461, time consumption: 0.086723
step: 15800, loss: 69.277306, time consumption: 0.084304
step: 15840, loss: 235.295761, time consumption: 0.089765
=> current epoch:  99
step: 15880, loss: 212.812363, time consumption: 0.129543
step: 15920, loss: 131.421616, time consumption: 0.114077
step: 15960, loss: 125.480614, time consumption: 0.102471
step: 16000, loss: 174.482361, time consumption: 0.101323
Processing a single batch of validation images at step: 16000
=> current epoch:  100
step: 16040, loss: 75.818176, time consumption: 0.104039
step: 16080, loss: 301.154144, time consumption: 0.102811
step: 16120, loss: 90.787491, time consumption: 0.094948
step: 16160, loss: 285.185303, time consumption: 0.097407
=> current epoch:  101
step: 16200, loss: 201.492462, time consumption: 0.091908
step: 16240, loss: 92.816666, time consumption: 0.086115
step: 16280, loss: 255.945175, time consumption: 0.087958
step: 16320, loss: 79.955139, time consumption: 0.090565
=> cu

step: 20880, loss: 108.115517, time consumption: 0.105805
step: 20920, loss: 182.229980, time consumption: 0.110890
step: 20960, loss: 134.058945, time consumption: 0.117851
=> current epoch:  131
step: 21000, loss: 139.373245, time consumption: 0.105542
step: 21040, loss: 115.686668, time consumption: 0.107328
step: 21080, loss: 156.371780, time consumption: 0.109497
step: 21120, loss: 113.993553, time consumption: 0.106197
=> current epoch:  132
step: 21160, loss: 80.084000, time consumption: 0.087054
step: 21200, loss: 181.077118, time consumption: 0.090325
step: 21240, loss: 83.571892, time consumption: 0.098667
step: 21280, loss: 106.640724, time consumption: 0.094480
=> current epoch:  133
step: 21320, loss: 75.455353, time consumption: 0.099327
step: 21360, loss: 240.006012, time consumption: 0.094681
step: 21400, loss: 241.251495, time consumption: 0.089615
step: 21440, loss: 88.776367, time consumption: 0.090108
=> current epoch:  134
step: 21480, loss: 278.520752, time consum

step: 26000, loss: 137.337784, time consumption: 0.082478
step: 26040, loss: 223.506882, time consumption: 0.082414
step: 26080, loss: 170.892151, time consumption: 0.087173
=> current epoch:  163
step: 26120, loss: 169.490753, time consumption: 0.124653
step: 26160, loss: 209.767731, time consumption: 0.115190
step: 26200, loss: 124.619240, time consumption: 0.106514
step: 26240, loss: 202.134628, time consumption: 0.100323
=> current epoch:  164
step: 26280, loss: 67.385597, time consumption: 0.123785
step: 26320, loss: 75.292297, time consumption: 0.137619
step: 26360, loss: 259.528839, time consumption: 0.136227
step: 26400, loss: 160.745514, time consumption: 0.135526
=> current epoch:  165
step: 26440, loss: 76.592949, time consumption: 0.086437
step: 26480, loss: 92.428001, time consumption: 0.100853
step: 26520, loss: 187.274429, time consumption: 0.102230
step: 26560, loss: 175.442612, time consumption: 0.101687
=> current epoch:  166
step: 26600, loss: 45.614845, time consump

step: 31240, loss: 176.055222, time consumption: 0.151310
step: 31280, loss: 86.265755, time consumption: 0.116177
step: 31320, loss: 116.314392, time consumption: 0.105909
step: 31360, loss: 145.714264, time consumption: 0.103918
=> current epoch:  196
step: 31400, loss: 214.062698, time consumption: 0.087585
step: 31440, loss: 109.635521, time consumption: 0.088234
step: 31480, loss: 141.963867, time consumption: 0.099601
step: 31520, loss: 291.051575, time consumption: 0.095088
=> current epoch:  197
step: 31560, loss: 112.016640, time consumption: 0.086740
step: 31600, loss: 142.583771, time consumption: 0.084711
step: 31640, loss: 72.134064, time consumption: 0.085274
step: 31680, loss: 178.966660, time consumption: 0.083695
=> current epoch:  198
step: 31720, loss: 98.734413, time consumption: 0.133544
step: 31760, loss: 126.709061, time consumption: 0.107668
step: 31800, loss: 94.021584, time consumption: 0.106160
step: 31840, loss: 135.801376, time consumption: 0.102858
=> curr

step: 36400, loss: 186.616882, time consumption: 0.086405
step: 36440, loss: 92.884323, time consumption: 0.086220
step: 36480, loss: 82.446304, time consumption: 0.086179
=> current epoch:  228
step: 36520, loss: 153.237839, time consumption: 0.103592
step: 36560, loss: 181.986923, time consumption: 0.116083
step: 36600, loss: 184.909042, time consumption: 0.110255
step: 36640, loss: 247.084579, time consumption: 0.102636
=> current epoch:  229
step: 36680, loss: 169.607056, time consumption: 0.114216
step: 36720, loss: 79.546227, time consumption: 0.109537
step: 36760, loss: 157.147095, time consumption: 0.107820
step: 36800, loss: 87.034470, time consumption: 0.104806
=> current epoch:  230
step: 36840, loss: 180.224426, time consumption: 0.139702
step: 36880, loss: 145.442841, time consumption: 0.117343
step: 36920, loss: 139.364014, time consumption: 0.104757
step: 36960, loss: 128.270859, time consumption: 0.099478
=> current epoch:  231
step: 37000, loss: 130.009567, time consum

=> current epoch:  260
step: 41640, loss: 173.834229, time consumption: 0.122512
step: 41680, loss: 267.696442, time consumption: 0.100742
step: 41720, loss: 69.242088, time consumption: 0.095073
step: 41760, loss: 77.380257, time consumption: 0.095797
=> current epoch:  261
step: 41800, loss: 131.968582, time consumption: 0.108738
step: 41840, loss: 57.033447, time consumption: 0.100594
step: 41880, loss: 249.457474, time consumption: 0.097231
step: 41920, loss: 102.488258, time consumption: 0.092719
=> current epoch:  262
step: 41960, loss: 213.114059, time consumption: 0.100588
step: 42000, loss: 106.385376, time consumption: 0.103277
step: 42040, loss: 62.056667, time consumption: 0.097843
step: 42080, loss: 112.739014, time consumption: 0.095455
=> current epoch:  263
step: 42120, loss: 75.311493, time consumption: 0.089857
step: 42160, loss: 139.262131, time consumption: 0.094439
step: 42200, loss: 141.539841, time consumption: 0.092448
step: 42240, loss: 136.544708, time consump

step: 46800, loss: 53.681770, time consumption: 0.107656
step: 46840, loss: 78.928848, time consumption: 0.102620
step: 46880, loss: 116.073570, time consumption: 0.096811
=> current epoch:  293
step: 46920, loss: 185.756119, time consumption: 0.086813
step: 46960, loss: 130.740341, time consumption: 0.116228
step: 47000, loss: 109.448914, time consumption: 0.107927
step: 47040, loss: 84.170464, time consumption: 0.111924
=> current epoch:  294
step: 47080, loss: 99.106621, time consumption: 0.084602
step: 47120, loss: 169.722824, time consumption: 0.115122
step: 47160, loss: 112.722557, time consumption: 0.106966
step: 47200, loss: 72.007217, time consumption: 0.111395
=> current epoch:  295
step: 47240, loss: 96.133247, time consumption: 0.126418
step: 47280, loss: 171.928070, time consumption: 0.115852
step: 47320, loss: 100.699043, time consumption: 0.106481
step: 47360, loss: 195.681870, time consumption: 0.099976
=> current epoch:  296
step: 47400, loss: 53.731625, time consumpti

step: 51960, loss: 58.935696, time consumption: 0.095477
step: 52000, loss: 144.415146, time consumption: 0.093118
=> current epoch:  325
step: 52040, loss: 450.796478, time consumption: 0.109327
step: 52080, loss: 134.461075, time consumption: 0.103504
step: 52120, loss: 265.690643, time consumption: 0.102918
step: 52160, loss: 178.588058, time consumption: 0.096912
=> current epoch:  326
step: 52200, loss: 176.099289, time consumption: 0.089228
step: 52240, loss: 151.714752, time consumption: 0.111258
step: 52280, loss: 74.905853, time consumption: 0.112013
step: 52320, loss: 95.450859, time consumption: 0.124803
=> current epoch:  327
step: 52360, loss: 156.190201, time consumption: 0.083458
step: 52400, loss: 220.809402, time consumption: 0.081524
step: 52440, loss: 248.933380, time consumption: 0.082329
step: 52480, loss: 37.510952, time consumption: 0.096061
=> current epoch:  328
step: 52520, loss: 256.866852, time consumption: 0.079262
step: 52560, loss: 96.669113, time consump

step: 57120, loss: 136.773407, time consumption: 0.093634
=> current epoch:  357
step: 57160, loss: 117.363846, time consumption: 0.083963
step: 57200, loss: 195.346512, time consumption: 0.088318
step: 57240, loss: 125.133911, time consumption: 0.096172
step: 57280, loss: 107.809258, time consumption: 0.095750
=> current epoch:  358
step: 57320, loss: 117.789604, time consumption: 0.094098
step: 57360, loss: 129.983398, time consumption: 0.101471
step: 57400, loss: 60.108936, time consumption: 0.106204
step: 57440, loss: 258.174286, time consumption: 0.099667
=> current epoch:  359
step: 57480, loss: 226.231689, time consumption: 0.102197
step: 57520, loss: 219.333359, time consumption: 0.100427
step: 57560, loss: 101.181168, time consumption: 0.096881
step: 57600, loss: 44.173504, time consumption: 0.098248
=> current epoch:  360
step: 57640, loss: 155.293442, time consumption: 0.097961
step: 57680, loss: 30.901716, time consumption: 0.091508
step: 57720, loss: 184.088791, time consu

step: 62280, loss: 182.345200, time consumption: 0.085645
step: 62320, loss: 86.946693, time consumption: 0.100590
step: 62360, loss: 154.563705, time consumption: 0.094897
step: 62400, loss: 80.731834, time consumption: 0.093382
=> current epoch:  390
step: 62440, loss: 150.359360, time consumption: 0.089076
step: 62480, loss: 223.492722, time consumption: 0.091480
step: 62520, loss: 79.990753, time consumption: 0.098601
step: 62560, loss: 127.052208, time consumption: 0.095056
=> current epoch:  391
step: 62600, loss: 147.544174, time consumption: 0.084798
step: 62640, loss: 66.242645, time consumption: 0.088252
step: 62680, loss: 145.337753, time consumption: 0.099058
step: 62720, loss: 79.033447, time consumption: 0.097444
=> current epoch:  392
step: 62760, loss: 297.152191, time consumption: 0.110667
step: 62800, loss: 83.447136, time consumption: 0.102881
step: 62840, loss: 157.006104, time consumption: 0.101922
step: 62880, loss: 42.821350, time consumption: 0.100936
=> current

step: 67440, loss: 68.373497, time consumption: 0.079601
step: 67480, loss: 164.208282, time consumption: 0.079558
step: 67520, loss: 230.260742, time consumption: 0.086900
=> current epoch:  422
step: 67560, loss: 154.150162, time consumption: 0.079282
step: 67600, loss: 155.866776, time consumption: 0.101107
step: 67640, loss: 99.302246, time consumption: 0.096112
step: 67680, loss: 126.852493, time consumption: 0.094783
=> current epoch:  423
step: 67720, loss: 94.343071, time consumption: 0.108619
step: 67760, loss: 106.359070, time consumption: 0.106296
step: 67800, loss: 130.032455, time consumption: 0.104168
step: 67840, loss: 101.981949, time consumption: 0.098236
=> current epoch:  424
step: 67880, loss: 116.088394, time consumption: 0.078515
step: 67920, loss: 41.765198, time consumption: 0.097374
step: 67960, loss: 198.920898, time consumption: 0.094180
step: 68000, loss: 123.438232, time consumption: 0.093045
=> current epoch:  425
step: 68040, loss: 180.554443, time consum

In [3]:
# 将所有测试集中的图片大小尺寸变为 16 的倍数
from PIL import Image, ImageOps
import os

def resize_to_multiple_of_sixteen(image_path, output_path):
    # 打开图片
    with Image.open(image_path) as img:
        width, height = img.size
        # 计算新的尺寸，保证宽度和高度都是16的倍数
        new_width = ((width + 15) // 16) * 16
        new_height = ((height + 15) // 16) * 16
        
        if new_width != width or new_height != height:
            # 如果尺寸需要改变，则进行调整
            resized_img = img.resize((new_width, new_height), resample=Image.Resampling.LANCZOS)
        else:
            # 尺寸已经是16的倍数，直接使用原图
            resized_img = img
        
        # 保存调整后的图片
        resized_img.save(output_path)

def process_images_in_directory(directory, output_dir):
    # 确保输出目录存在
    if not os.path.exists(output_dir):
        os.makedirs(output_dir)
    
    for filename in os.listdir(directory):
        if filename.lower().endswith(('.png', '.jpg', '.jpeg', '.bmp', '.gif')):
            file_path = os.path.join(directory, filename)
            out_file_path = os.path.join(output_dir, filename)
            resize_to_multiple_of_sixteen(file_path, out_file_path)
            print(f"Processed {filename}")

# 使用示例
input_directory = './data/raindrop/test/target'  # 输入图片所在目录
output_directory = './data/raindrop/test/target'  # 输出图片存储目录
process_images_in_directory(input_directory, output_directory)

Processed 0_clean.jpg
Processed 1_clean.jpg
Processed 10_clean.jpg
Processed 100_clean.jpg
Processed 101_clean.jpg
Processed 102_clean.jpg
Processed 103_clean.jpg
Processed 104_clean.jpg
Processed 105_clean.jpg
Processed 106_clean.jpg
Processed 107_clean.jpg
Processed 108_clean.jpg
Processed 109_clean.jpg
Processed 11_clean.jpg
Processed 110_clean.jpg
Processed 111_clean.jpg
Processed 112_clean.jpg
Processed 113_clean.jpg
Processed 114_clean.jpg
Processed 115_clean.jpg
Processed 116_clean.jpg
Processed 117_clean.jpg
Processed 118_clean.jpg
Processed 119_clean.jpg
Processed 12_clean.jpg
Processed 120_clean.jpg
Processed 121_clean.jpg
Processed 122_clean.jpg
Processed 123_clean.jpg
Processed 124_clean.jpg
Processed 125_clean.jpg
Processed 126_clean.jpg
Processed 127_clean.jpg
Processed 128_clean.jpg
Processed 129_clean.jpg
Processed 13_clean.jpg
Processed 130_clean.jpg
Processed 131_clean.jpg
Processed 132_clean.jpg
Processed 133_clean.jpg
Processed 134_clean.jpg
Processed 135_clean.jpg


测试模型（raindrop的test数据集）（预训练的模型(epoch 600, step 96000)）（input、target）(249张图片)

In [4]:
!python eval_diffusion.py --config ./configs/raindrop_configs.yml

=> using device: cuda
=> creating diffusion model
=> loaded checkpoint './pretrained_model/.pth.tar' (epoch 600, step 96000)
=> starting processing image named ['0_rain']
=> starting processing image named ['1_rain']
=> starting processing image named ['10_rain']
=> starting processing image named ['100_rain']
=> starting processing image named ['101_rain']
=> starting processing image named ['102_rain']
=> starting processing image named ['103_rain']
=> starting processing image named ['104_rain']
=> starting processing image named ['105_rain']
=> starting processing image named ['106_rain']
=> starting processing image named ['107_rain']
=> starting processing image named ['108_rain']
=> starting processing image named ['109_rain']
=> starting processing image named ['11_rain']
=> starting processing image named ['110_rain']
=> starting processing image named ['111_rain']
=> starting processing image named ['112_rain']
=> starting processing image named ['113_rain']
=> starting proce

=> starting processing image named ['27_rain']
=> starting processing image named ['28_rain']
=> starting processing image named ['29_rain']
=> starting processing image named ['3_rain']
=> starting processing image named ['30_rain']
=> starting processing image named ['31_rain']
=> starting processing image named ['32_rain']
=> starting processing image named ['33_rain']
=> starting processing image named ['34_rain']
=> starting processing image named ['35_rain']
=> starting processing image named ['36_rain']
=> starting processing image named ['37_rain']
=> starting processing image named ['38_rain']
=> starting processing image named ['39_rain']
=> starting processing image named ['4_rain']
=> starting processing image named ['40_rain']
=> starting processing image named ['41_rain']
=> starting processing image named ['42_rain']
=> starting processing image named ['43_rain']
=> starting processing image named ['44_rain']
=> starting processing image named ['45_rain']
=> starting pro

评估模型（计算 PSNR 和 SSIM ）

In [5]:
!python calculate_psnr_ssim.py --config ./configs/raindrop_configs.yml

Processing image: 0_rain.png
PSNR is 22.7829 and SSIM is 0.8316
Processing image: 100_rain.png
PSNR is 22.9989 and SSIM is 0.8314
Processing image: 101_rain.png
PSNR is 18.7596 and SSIM is 0.7623
Processing image: 102_rain.png
PSNR is 25.1966 and SSIM is 0.8968
Processing image: 103_rain.png
PSNR is 25.4550 and SSIM is 0.8607
Processing image: 104_rain.png
PSNR is 19.5342 and SSIM is 0.8720
Processing image: 105_rain.png
PSNR is 23.3016 and SSIM is 0.7825
Processing image: 106_rain.png
PSNR is 19.9584 and SSIM is 0.5270
Processing image: 107_rain.png
PSNR is 24.4373 and SSIM is 0.8824
Processing image: 108_rain.png
PSNR is 19.6618 and SSIM is 0.5317
Processing image: 109_rain.png
PSNR is 16.1630 and SSIM is 0.3121
Processing image: 10_rain.png
PSNR is 25.1070 and SSIM is 0.8853
Processing image: 110_rain.png
PSNR is 23.1448 and SSIM is 0.7873
Processing image: 111_rain.png
PSNR is 22.0851 and SSIM is 0.6876
Processing image: 112_rain.png
PSNR is 21.5616 and SSIM is 0.8443
Processing im

PSNR is 23.7978 and SSIM is 0.9099
Processing image: 212_rain.png
PSNR is 25.8524 and SSIM is 0.8648
Processing image: 213_rain.png
PSNR is 27.8141 and SSIM is 0.9267
Processing image: 214_rain.png
PSNR is 22.8507 and SSIM is 0.8698
Processing image: 215_rain.png
PSNR is 25.6597 and SSIM is 0.8820
Processing image: 216_rain.png
PSNR is 23.7676 and SSIM is 0.8268
Processing image: 217_rain.png
PSNR is 25.7680 and SSIM is 0.8394
Processing image: 218_rain.png
PSNR is 22.4558 and SSIM is 0.8264
Processing image: 219_rain.png
PSNR is 23.9081 and SSIM is 0.8797
Processing image: 21_rain.png
PSNR is 20.1527 and SSIM is 0.6715
Processing image: 220_rain.png
PSNR is 20.6388 and SSIM is 0.7805
Processing image: 221_rain.png
PSNR is 26.2021 and SSIM is 0.8893
Processing image: 222_rain.png
PSNR is 18.2089 and SSIM is 0.4225
Processing image: 223_rain.png
PSNR is 19.5172 and SSIM is 0.6129
Processing image: 224_rain.png
PSNR is 25.2220 and SSIM is 0.8295
Processing image: 225_rain.png
PSNR is 24.

最终模型的效果可看：Testing set, PSNR is 22.6955 and SSIM is 0.7739